rct229/rulesets/ashrae9012019/section19/section19rule21.py

```diff
@@ -0,0 +1,510 @@
+
+    def create_data(self, context, data):
+        rmi_b = context.baseline
+        rmi_p = context.proposed
+        climate_zone = rmi_b["weather"]["climate_zone"]
+
+        ER_not_req_for_heating_sys = (
+            True if climate_zone in APPLICABLE_CZ_ZONES else False
+        )
+
+        dict_of_zones_and_terminal_units_served_by_hvac_sys = (
+            get_dict_of_zones_and_terminal_units_served_by_hvac_sys(rmi_b)
+        )
+
+        hvac_sys_and_assoc_zones_largest_exhaust_source = {}
+        for hvac_id_b in find_all(
+            "$.buildings[*].building_segments[*].heating_ventilating_air_conditioning_systems[*]",
+            rmi_b,
+        ):
+            for zone_id_b in dict_of_zones_and_terminal_units_served_by_hvac_sys[
+                hvac_id_b
+            ]["zone_list"]:
+                space_p = find_one(
+                    f"$.buildings[*].building_segments[*].zones[?(@.id == {zone_id_b})].spaces",
+                    rmi_p,
+                )
+                if space_p.get("lighting_space_type") is not None:
+                    lighting_space_type_p = space_p["lighting_space_type"]
+                    serves_zones_with_systems_likely_exhausting_toxic_etc = (
+                        True
+                        if lighting_space_type_p
+                        == LIGHTING_SPACE.LABORATORY_EXCEPT_IN_OR_AS_A_CLASSROOM
+                        else False
+                    )
+                    serves_kitchen_space = (
+                        True
+                        if lighting_space_type_p == LIGHTING_SPACE.FOOD_PREPARATION_AREA
+                        else False
+                    )
+                else:
+                    all_lighting_space_types_defined = False
+
+                if space_p.get("ventilation_space_type") is not None:
+                    ventilation_space_type_p = space_p["ventilation_space_type"]
+                    serves_zones_with_systems_likely_exhausting_toxic_etc = (
+                        True
+                        if ventilation_space_type_p
+                        == LIGHTING_SPACE.MISCELLANEOUS_SPACES_MANUFACTURING_WHERE_HAZARDOUS_MATERIALS_ARE_USED_EXCLUDES_HEAVY_INDUSTRIAL_AND_CHEMICAL_PROCESSES
+                        else False
+                    )
+
+                    serves_kitchen_space = (
+                        True
+                        if ventilation_space_type_p
+                        == LIGHTING_SPACE.FOOD_AND_BEVERAGE_SERVICE_KITCHEN_COOKING
+                        else False
+                    )
+                else:
+                    all_ventilation_space_types_defined = False
+
+            hvac_sys_and_assoc_zones_largest_exhaust_source[
+                hvac_id_b
+            ] = get_hvac_sys_and_assoc_zones_largest_exhaust_source(rmi_b, hvac_id_b)
+
+        for hvac_id_p in get_list_hvac_systems_associated_with_zone(rmi_p, zone_id_b):
+            hvac_p = find_one(
```

Comment:
It is hard to see if this loop is inside of zone loop or not.
However, even if it is inside, I will suggest to move it outside for two reasons:
1. Clarify, we will just focus on processing data from proposed HVAC system in one loop
2. Performance, reduce O3 computation time to O2.

Reply:
I agree. However, we need to provide ```zone_id_b``` to the ```get_list_hvac_systems_associated_with_zone``` function. So, can't we avoid the three nested loops?